modules/biobank/php/containercontroller.class.inc

```diff
@@ -0,0 +1,403 @@
+    {
+        switch($action) {
+        case 'GET':
+            if (!$this->user->hasPermission('biobank_container_view')) {
+                throw new \Forbidden('Container: View Permission Denied');
+            }
+        case 'PUT':
+            if (!$this->user->hasPermission('biobank_container_update')) {
+                throw new \Forbidden('Container: Update Permission Denied');
+            }
+        case 'POST':
+            if (!$this->user->hasPermission('biobank_container_create')) {
+                throw new \Forbidden('Container: Create Permission Denied');
+            }
+        }
+    }
+
+    /**
+     * Validates Container Object before being saved
+     *
+     * @param Container $container Container to be validated
+     *
+     * @return void
+     */
+    public function validateInstance(Container $container) : void
+    {
+        $this->validateBarcode($container);
+        $this->validateTypeId($container);
+        $this->validateTemperature($container);
+        $this->validateStatusId($container);
+        $this->validateCenterId($container);
+        $this->validateParentContainerId($container);
+        $this->validateCoordinate($container);
+    }
+
+    /**
+     * Validates Container Object Barcode
+     *
+     * @param Container $container Container to be checked.
+     *
+     * @throws BadRequest if the provided Container does not meet validation requirements
+     *
+     * @return void
+     */
+    private function validateBarcode(Container $container) : void
+    {
+        if (is_null($container->getBarcode())) {
+            throw new \BadRequest('Barcode value must be set');
+        }
+
+        if (!isset($container->getId())) {
+            $containerList = $this->dao->selectInstances();
+            foreach ($containerList as $c) {
+                $b = $c->getBarcode();
+                if ($b === $barcode) {
+                    throw new \BadRequest('Barcode must be unique');
+                }
+            }
+        }
+
+        //TODO: Regex check on Barcode.
+        //i.e.
+        //$specimenTypes = $this->dao->getSpecimenTypes();
+        //$regex = $specimenTypes[$specimen['typeId']]['regex'];
+        //if (isset($regex)) {
+        //    if (preg_match($regex, $container['barcode']) !== 1) {
+        //        throw new \BadRequest('Barcode is not of proper format for the
+        //                               selected specimen type');
+        //    }
+        //}
+    }
+
+    /**
+     * Validates Container Object Type ID.
+     *
+     * @param Container $container Container to be checked.
+     *
+     * @throws BadRequest if the provided Container does not meet validation requirements
+     *
+     * @return void
+     */
+    private function validateTypeId(Container $container) : void
+    {
+        if (is_null($container->getTypeId())) {
+            throw new \BadRequest('Type value must be set');
+        }
```

Comment:
Are there any cases where it should be valid to have a Container with no type id? (And if not, why is the TypeId nullable? It's better to make invalid states unrepresentable.)

Reply:
@HenriRabalais This is what I was talking about earlier also, not really sure where it makes sense for it to not have an ID.

Reply:
This is a `typeID`, not an `id`. So the `id` is nullable because a container only receives on ID once it's inserted in the database. On the other hand, a container necessitates a typeID to be inserted into the database.

Reply:
`ContainerEndpoint::handlePOST($request)` could check if the parameter typeId is null up front. Also, `new ContainerTypeById($typeId)` could throw an exception ... but this should be discuss as mentioned in an other comment ;)